nomos-libp2p/src/behaviour/nat/state_machine.rs

```diff
@@ -0,0 +1,749 @@
+                    addr,
+                    self.state.addr(),
+                );
+                self
+            }
+            _ => self,
+        }
+    }
+}
+
+impl OnEvent<Event> for State<Private, Event> {
+    fn on_event(self: Box<Self>, event: Event, _: &CommandTx) -> Box<dyn OnEvent<Event>> {
+        match event {
+            Event::LocalAddressChanged(addr) => self.boxed(|state| state.into_test_if_public(addr)),
+            Event::DefaultGatewayChanged(_) => {
+                let local_addr = self.state.addr().clone();
+                self.boxed(|state| state.into_test_if_public(local_addr))
+            }
+            _ => self,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashSet;
+
+    use fixtures::*;
+    use tokio::sync::mpsc::{error::TryRecvError, unbounded_channel};
+
+    use super::*;
+
+    #[cfg(test)]
+    impl<S: PartialEq + 'static, E: PartialEq + 'static> DynPartialEq for State<S, E> {
+        fn as_any(&self) -> &dyn core::any::Any {
+            self
+        }
+
+        fn box_eq(&self, other: &dyn core::any::Any) -> bool {
+            other.downcast_ref::<Self>() == Some(self)
+        }
+    }
+
+    #[cfg(test)]
+    impl<E: PartialEq> PartialEq for Box<dyn OnEvent<E>> {
+        fn eq(&self, other: &Self) -> bool {
+            self.box_eq(other.as_any())
+        }
+    }
+
+    #[cfg(test)]
+    impl<E: PartialEq> PartialEq<&Self> for Box<dyn OnEvent<E>> {
+        fn eq(&self, other: &&Self) -> bool {
+            self.box_eq(other.as_any())
+        }
+    }
+
+    fn all_events<'a>() -> HashSet<TestEvent<'a>> {
+        [
+            autonat_failed(),
+            autonat_failed_with_addr_mismatch(),
+            autonat_ok(),
+            autonat_ok_with_address_mismatch(),
+            mapping_failed(),
+            mapping_failed_with_addr_mismatch(),
+            mapping_ok(),
+            new_external_address_candidate(),
+            external_address_confirmed(),
+            external_address_confirmed_with_addr_mismatch(),
+            default_gateway_changed(),
+            local_address_changed(),
+            other_from_swarm_event(),
+        ]
+        .into()
+    }
+
+    fn init_state_machine(
+        init_state: Box<dyn OnEvent<Event>>,
+        command_tx: UnboundedSender<Command>,
+    ) -> StateMachine {
+        StateMachine {
+            inner: Some(init_state),
+            command_tx: command_tx.into(),
+        }
+    }
+
+    type TestCase<'a> = Vec<(
+        u32,
+        Box<dyn Fn() -> Box<dyn OnEvent<Event>>>,
+        Vec<(
+            TestEvent<'a>,
+            Box<dyn Fn() -> Box<dyn OnEvent<Event>>>,
+            Option<Command>,
+        )>,
+    )>;
+
+    #[expect(clippy::too_many_lines, reason = "Keep the test readable")]
+    #[test]
+    fn test_transitions_and_emitted_commands() {
+        let expected_transitions: TestCase<'_> = vec![
+            (
+                line!(),
```

Comment:
Question: why not using a string identifier instead, independently of the line number?